### ISSUE TITLE
Simplification Error

### ISSUE THREAD

**Issue by srush** [google-research/dex-lang#612]
Getting this error after updating. Think it is a regression from a recent update. 

Can't figure out what code is causing it. Think it might be parser? 

```
Unexpected exception-throwing expression: (%runWriter
   '\SOH'
   (\tmp268:Word8 tmp269:Word8. %and tmp268 tmp269)
   (\h:Type.
      \ref:(Ref
              h
              Word8) ->{Accum h, State h', State h'1, State h'2, State h'3, Except}
        for tmp268:(Fin 2).
          tmp269:Word8 = tmp267 tmp268
          extend ref using \tmp270:Word8. %and tmp269 tmp270))
CallStack (from HasCallStack):
  error, called at src/lib/Simplify.hs:568:11 in dex-0.1.0.0-cplxjX7IqYJJeaxgDLDjq:Simplify
```

**Comment by apaszke** [comment]
Nothing comes to mind just from the error message. Is there a small program we could use to reproduce this?

**Comment by srush** [comment]
```
import parser                                                                                                                                                                                                        
                                                                                                                                                                                                                     
def parseList (parser:Parser a) : Parser (List a) = MkParser \h.                                                                                                                                                     
  yieldState (AsList _ []) \results.                                                                                                                                                                                 
    iter \_.                                                                                                                                                                                                         
      maybeVal = parse h $ optional parser                                                                                                                                                                           
      case maybeVal of                                                                                                                                                                                               
        Nothing -> Done ()                                                                                                                                                                                           
        Just x ->                                                                                                                                                                                                    
          push results x                                                                                                                                                                                             
          Continue                                                                                                                                                                                                   
                                                                                                                                                                                                                     
def parseChars (n:Type) : (Parser (n => Char)) =                                                                                                                                                                     
    MkParser \h. for i. parse h parseAny                                                                                                                                                                             
                                                                                                                                                                                                                     
runParserPartial "ab" $ parseList $ parseChars (Fin 2) 
```

**Comment by srush** [comment]
Oh actually here is a much smaller one 

```haskell
import parser                                                                               
runParser "a" $ (MkParser \h. Just (parse h parseInt)) <|> return Nothing
```
Seems maybe related to the `catch` statement in `<|>` which seems to not be working. 

```haskell
def (<|>) (p1:Parser a) (p2:Parser a) : Parser a = MkParser \h.                             
  (s, posRef) = h                                                                           
  curPos = get posRef                                                                       
  case catch do parse h p1 of                                                               
    Nothing ->                                                                              
      assert $ curPos == get posRef                                                         
      parse h p2                                                                            
    Just ans -> ans  
```

```
import parser
dex: Unexpected exception-throwing expression: (%runWriter
   '\SOH'
   (\tmp60:Word8 tmp61:Word8. %and tmp60 tmp61)
   (\h4:Type.
      \ref4:(Ref
               h4
               Word8) ->{Accum h4, State h, State h1, State h2, State h', Except}
        for tmp60:(Fin tmp57).
          tmp61:Word8 = tmp59 tmp60
          extend ref4 using \tmp62:Word8. %and tmp61 tmp62))
CallStack (from HasCallStack):
  error, called at src/lib/Simplify.hs:568:11 in dex-0.1.0.0-cplxjX7IqYJJeaxgDLDjq:Simplify
```

**Comment by srush** [comment]
One more attempt to reduce this without Parser. Super weird bug. 

Fails: 
```haskell
def f (h: Unit) : {Except} List Int =                                                       
  yieldState mempty \results.                                                               
      results := (get results) <> AsList 1 [2]                                              
                                                                                            
catch do (catch do f ())  
```

Works fine
```haskell
def f (h: Unit) : {Except} List Int =                                                       
  yieldState mempty \results.                                                               
      results :=  AsList 1 [2]                                              
                                                                                            
catch do (catch do f ())  
```

**Comment by dougalm** [comment]
Thanks for the easy repro! It looks like I introduced the regression in baf5cfbe9af01818277de3970998291c2acc1943 when I switched to accumulating exceptions using the writer effect instead of state. It's good for parallelization but it turns out we hadn't implemented exception handling for the writer effect, so it breaks some cases of nested `catch`, like your example. The fix is in #615!
